preprocess/docenc_main.cc

```diff
@@ -0,0 +1,258 @@
+#include <algorithm>
+#include <cstdio>
+#include <cstdlib>
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+#include <unistd.h>
+#include <vector>
+#include "util/file_stream.hh"
+#include "util/file_piece.hh"
+#include "util/string_piece.hh"
+#include "preprocess/base64.hh"
+
+using namespace std;
+
+enum Mode {
+	COMPRESS,
+	DECOMPRESS
+};
+
+string prefix_lines(string const &input, string const prefix)
+{
+	stringstream sin(input);
+	stringstream sout;
+
+	string line;
+	while (getline(sin, line))
+		sout << prefix << line << '\n';
+
+	return sout.str();
+}
+
+size_t decode(util::FilePiece &in, util::FileStream &out, char delimiter, vector<size_t> const &indices, bool print_document_index, bool &delimiter_encountered) {
+	size_t document_index = 0;
+	vector<size_t>::const_iterator indices_it(indices.begin());
+
+
+	for (StringPiece line : in) {
+		++document_index;
+
+		if (!indices.empty()) {
+			if (*indices_it != document_index) {
+				continue; // skip document
+			} else {
+				indices_it++;
+			}
+		}
+
+		string document;
+		preprocess::base64_decode(line, document);
+
+		if (!delimiter_encountered && document.find(delimiter == '\n' ? string("\n\n") : string(&delimiter, 1)) != string::npos)
+			delimiter_encountered = true;
+
+		if (print_document_index)
+			document = prefix_lines(document, to_string(document_index) + "\t");
+
+		out << document << delimiter;
+
+		// Have we found all our indices? Then stop early
+		if (!indices.empty() && indices_it == indices.end())
+			break;
+	}
+
+	return document_index;
+}
+
+size_t encode(util::FilePiece &in, util::FileStream &out, char delimiter, vector<size_t> const &indices) {
+	size_t document_index = 0;
+	string document;
+	vector<size_t>::const_iterator indices_it(indices.begin());
+
+	bool is_eof = false;
+	while (!is_eof) {
+		document.clear();
+		
+		// Start accumulating lines that make up a document
+		StringPiece line;
+		while (true) {
+			is_eof = !in.ReadLineOrEOF(line, delimiter, true);
+			
+			if (is_eof)
```

Comment:
while (in.ReadLineOrEOF(line, delimiter, true))

Reply:
There are more cases that can set is_eof to true, and there are more loops/clauses that check whether is_eof is true.